lib/config/presets/index.ts

```diff
@@ -340,15 +363,67 @@ export async function resolveConfigPresets(
         val as RenovateConfig,
         baseConfig,
         ignorePresets,
-        existingPresets
+        existingPresets,
+        shallowResolve
       );
     }
   }
   logger.trace({ config: inputConfig }, 'Input config');
   logger.trace({ config }, 'Resolved config');
+  if (shallowResolve) {
+    handleExtendsArray(config, unresolvedPresets);
+  }
   return config;
 }
 
+// handleExtendsArray:
+// This function edits extends array presets when shallowResolve is enabled
+// 1. remove resolved external presets from the extends array.
+// 2. clean duplicate presets keys (from different resolution levels).
+function handleExtendsArray(
+  config: AllConfig,
+  unresolvedPresets: string[]
+): void {
+  if (!config.extends?.length) {
+    return;
+  }
+  const filteredPresets = config.extends.filter(
+    (e) => !shouldShallowResolve(e)
+  );
+  const uniqueExtends = new Set([...filteredPresets, ...unresolvedPresets]);
+  config.extends = Array.from(uniqueExtends);
+  if (config?.extends?.length === 0) {
+    // clean empty extends array
+    delete config.extends;
+  }
+}
+
+export function shouldShallowResolve(presetSource: string): boolean {
```

Comment:
```suggestion
export function skipDuringShallowResolve(presetSource: string): boolean {
```

This function is called three times and each time with `!` before it. Therefore it would be more readable to flip the naming/logic so that it returns true if it should skip and false if it should resolve. This means changing the true/false logic within the function and removing the `!` from the three locations